accepted/2023/simplify-output-paths.md

```diff
@@ -0,0 +1,130 @@
+
+```xml
+<Project>
+    <PropertyGroup>
+        <ArtifactsPath>$(MSBuildThisFileDirectory)\output</ArtifactsPath>
+    </PropertyGroup>
+</Project>
+```
+
+Under the artifacts folder will be nested folders for the type of the output, the project name, and the pivots, ie `<ArtifactsPath>\<Type of Output>\<Project Name>\<Pivots>`.
+
+- Type of Output - This will be a value such as `bin`, `publish`, `intermediates`, or `package`.  Projects will be able to override this value, for example to separate shipping and non-shipping artifacts into different folders
+  - `bin` will be the folder for the normal build output.  Projects can override this with `ArtifactsOutputName`
+  - `publish` will be the folder for the publish output.  Projects can override this with `ArtifactsPublishOutputName`
+  - `package` will be the folder for the package output, where .nupkg folders are placed. (Or should it be `packages`?  Feedback welcome.)  Projects can override this with `ArtifactsPackageOutputName`
+  - `intermediates` will be the folder for the intermediate output.  (Or should we stick with `obj`?  Feedback welcome.)
+- The Project Name.  This will ensure that each project has a separate output folder.  By default this will be `$(MSBuildProjectName)`, but projects can override this with the `ArtifactsProjectName` property.  If `ArtifactsPath` was not explicitly specified, and neither a `Directory.Build.props` nor a `.sln` file was found, then the artifacts folder will already be inside the project folder, and an additional Project Name folder will *not* be included in the output paths.  Additionally, for package output, this folder won't be included in the path, so that all the `.nupkg` files that are built can be in a single folder.
+- Pivots - This is used to distinguish between builds of a project for different configurations, target frameworks, runtime identifiers, or other values.  It can be specified with `ArtifactsPivots`.  By default, the pivot will include:
+  - The `Configuration` value (lower-cased, using the invariant culture)
+  - The `TargetFramework` if the project is multi-targeted
+  - The `RuntimeIdentifier`, if it was explicitly set (either in the project file or on the command line).  The `RuntimeIdentifier` won't be appended if it was automatically set by the SDK, for example because `SelfContained` was set
+
+If multiple pivot elements are used, they will be joined by the underscore (`_`) character.  For example, a multi-targeted project with the RuntimeIdentifer specified could have a pivot of `debug_net8.0-windows10.0.19041.0_win-x64`. For the `package` output type, the pivots will only be the `Configuration`, and other values won't be included in the path
+
+Some examples:
+
+- `.artifacts\bin\debug` - The build output path for a simple project when you run `dotnet build`
```

Comment:
What constitutes a "simple" project? I'm curious what happens when the project evolves such that the project name needs to be added (i.e. the `.artifacts\bin\MyApp\debug_net8.0` example below). Can you end up with the outputs of `MyApp` in both folders?

Reply:
Basically a non-multitargeted project, without a RuntimeIdentifier specified.  Yes, if you switch from single to multitargeted, there will be outputs that won't be cleaned up.  It's slightly better than what we have today, where if you retarget a single-targeted project (ie to update it from .NET 6 to .NET 7) you'll end up with extra folders.